Sources/BigInt/BigInt.swift

```diff
@@ -0,0 +1,810 @@
+          return true
+        }
+      }
+    }
+
+    return false
+  }
+}
+
+extension BigInt: LosslessStringConvertible {
+
+  public init?(_ description: String) {
+    self.init(description, radix: 10)
+  }
+
+  public init?<T>(_ description: T, radix: Int = 10) where T: StringProtocol {
+    precondition(2 ... 36 ~= radix, "Radix not in range 2 ... 36")
+
+    self = 0
+
+    let isNegative = description.hasPrefix("-")
+    let hasPrefix = isNegative || description.hasPrefix("+")
+    let utf8 = description.utf8.dropFirst(hasPrefix ? 1 : 0)
+    guard !utf8.isEmpty else { return nil }
+
+    for var byte in utf8 {
+      switch byte {
+      case UInt8(ascii: "0") ... UInt8(ascii: "9"):
+        byte -= UInt8(ascii: "0")
+      case UInt8(ascii: "A") ... UInt8(ascii: "Z"):
+        byte -= UInt8(ascii: "A")
+        byte += 10
+      case UInt8(ascii: "a") ... UInt8(ascii: "z"):
+        byte -= UInt8(ascii: "a")
+        byte += 10
+      default:
+        return nil
+      }
+      guard byte < radix else { return nil }
+      self *= BigInt._digits[radix]
+      self += BigInt._digits[Int(byte)]
+    }
+
+    if isNegative {
+      self.negate()
+    }
+  }
+}
+
+extension BigInt: Decodable {
+
+  public init(from decoder: Decoder) throws {
+    let singleValueContainer = try decoder.singleValueContainer()
+    let description = try singleValueContainer.decode(String.self)
+    guard let result = BigInt(description) else {
+      throw DecodingError.dataCorruptedError(
+        in: singleValueContainer,
+        debugDescription: "BigInt(\(description.debugDescription)) failed")
+    }
+    self = result
+  }
+}
+
+extension BigInt: Encodable {
+
+  public func encode(to encoder: Encoder) throws {
+    var singleValueContainer = encoder.singleValueContainer()
+    try singleValueContainer.encode(description)
+  }
+}
+
+// MARK: - Numeric Protocols
+
+extension BigInt: ExpressibleByIntegerLiteral {
+
+  public init(integerLiteral value: Int) {
+    if value >= 0, value < BigInt._digits.count {
+      self = BigInt._digits[value]
+    } else {
+      words = [UInt(bitPattern: value)]
+    }
+  }
+}
+
+extension BigInt: AdditiveArithmetic {
+
+  @inlinable
+  public static func + (lhs: BigInt, rhs: BigInt) -> BigInt {
+    var result = lhs
+    result += rhs
+    return result
+  }
+
+  public static func += (lhs: inout BigInt, rhs: BigInt) {
+    if lhs.words.count == 1, rhs.words.count == 1 {
+      let lhsWord = lhs.words[0]
+      let rhsWord = rhs.words[0]
+
+      let (result, isOverflow) = lhsWord.addingReportingOverflow(rhsWord)
+
+      if !isOverflow && result < Int.max {
+        lhs.words[0] = result
+        return
+      }
+      let knownNegativeResult = lhsWord > Int.max && rhsWord > Int.max
+
+      if lhsWord > Int.max || rhsWord > Int.max, !knownNegativeResult {
+        // positive + negative is always smaller, so overflow is a red herring
+        lhs.words[0] = result
+        return
```

Comment:
Have you evaluated using pattern matches to clarify some of this code and make it a bit easier to spot any edge cases. For instance, here I believe you could do:

```swift

    if lhs.words.count == 1, rhs.words.count == 1 {
      let lhsWord = lhs.words[0]
      let rhsWord = rhs.words[0]
      let lhsNegative = lhsWord > Int.max
      let rhsNegative = rhsWord > Int.max
      let sum = lhsWord &+ rhsWord
      let sumNegative = sum > Int.max
      switch (lhsNegative, rhsNegative, sumNegative) {
        case (true, true, false);
          lhs.words = [sum, UInt.max]
          return
        case (false, false, true);
          lhs.words = [sum, 0]
          return
        default:
          lhs.words[0] = sum
          return
      }
    }
```